### ISSUE TITLE
Add `alias` as a `field()` parameter for dataclasses

### ISSUE THREAD

**Issue by tmke8** [python/cpython#101192]
# Feature or enhancement

Add an `alias` parameter to `dataclasses.field()` which behaves like `alias` in `attrs.field()` (see [here](https://www.attrs.org/en/stable/api.html#attr.ib) and search for "alias"). It has the effect that it changes the corresponding argument name in the synthesized `__init__`.

# Pitch
Since attrs 22.2.0, `attrs.field` has an `alias` parameter that does the following:

```python
from attrs import define, field

@define
class C:
    _name: tuple[str, str] = field(alias="name")

    @property
    def full_name(self) -> str:
        return " ".join(self._name)

c = C(name=("John", "Smith"))  # `name` not `_name`
print(c.full_name)
```
The typical use-case for this is private instance variables as in the above example. But there could be other use-cases as well, like
```python
@define
class C:
    viewport: tuple[int, int] = field(alias="vp")

C(vp=(2, 4))  # constructor uses shorthand
```

Besides being a useful feature in and of itself, another reason to add this to the standard library is that it would bring dataclasses in line with `dataclass_transform` which already supports the `alias` parameter in `field()`: https://peps.python.org/pep-0681/#field-specifier-parameters . It is currently a bit strange that `dataclass_transform` has a feature that `dataclasses.dataclass` doesn't actually have.

# Previous discussion

https://discuss.python.org/t/add-alias-as-a-field-parameter-for-dataclasses/22988/1

**Comment by ericvsmith** [comment]
I’ve got this partially working. I’ll investigate attrs to make sure we’re in sync. Once I get a PR, hopefully some typing folks can take a look.

**Comment by vvzen** [comment]
Another use case could be to allow doing breaking changes when one wants to rename the spelling of a property.
For example, given a class like this:
```python
@dataclass
class SubmissionData(object):
    allowlist: list[str]
```

If I wanted to rename the spelling of `allowlist` to be `allow_list` without breaking existing client code, I could just alias `allowlist` to `allow_list` , say like this:

```python
@dataclass
class SubmissionData(object):
    allow_list: list[str] = field(alias="allowlist")
```

 Then, only in a major version of my package, I would drop support `allowlist` completely, etc.
 Thanks!

**Comment by tmke8** [comment]
Note though that when you define an `alias` only the alias name will be available in the `__init__`, so people would not be able to use the new name `allow_list` in the constructor.

**Comment by padrepitufo** [comment]
that seems to break the definition of "alias" for me.  I wonder if we really have to be 1:1 with attrs.

**Comment by tmke8** [comment]
Hmm, my main use case would be private attributes where you wouldn't want to use the non-alias name in the constructor:
```python
@dataclass
class C:
    _name: str = field(alias="name")
```
but I guess there might be other use cases. Though if you want to keep both names in the constructor, it would get a bit complicated. You'd need code for dealing with the case where both names are used at the same time.

I agree though that the word "alias" sounds like both names will be available. Unfortunately it's already enshrined in PEP 681...

**Comment by GalaxySnail** [comment]
There is a workaround with mypy. For details: https://web.archive.org/web/20240109161017/https://florimond.dev/en/posts/2018/10/reconciling-dataclasses-and-properties-in-python

tl;dr:

``` python
from dataclasses import dataclass, field

@dataclass
class Vehicle:

    foo: str
    wheels: int
    _wheels: int = field(init=False, repr=False)

    @property  # type: ignore[no-redef]
    def wheels(self) -> int:
        print("getting wheels")
        return self._wheels

    @wheels.setter
    def wheels(self, wheels: int) -> None:
        print("setting wheels to", wheels)
        self._wheels = wheels
```

It works well with `dataclass.replace`:

``` python
>>> a = Vehicle("bar", 42)
setting wheels to 42
>>> print(a)
getting wheels
Vehicle(foo='bar', wheels=42)

>>> from dataclasses import replace

>>> b = replace(a, wheels=12345)
setting wheels to 12345
>>> print(b)
getting wheels
Vehicle(foo='bar', wheels=12345)

>>> c = replace(a, foo="baz")
getting wheels
setting wheels to 42
>>> print(c)
getting wheels
Vehicle(foo='baz', wheels=42)
```

**Comment by Dobatymo** [comment]
Pydantic allows alternative names via alias in fields https://docs.pydantic.dev/latest/concepts/alias/
